tests/Test.ADAL.NET.Integration/ManagedUsercredentialTests.cs

```diff
@@ -69,5 +72,212 @@ public void Initialize()
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content =
+            new StringContent("{\"ver\":\"1.0\",\"account_type\":\"Managed\",\"domain_name\":\"id.com\"}")
+                },
+                QueryParams = new Dictionary<string, string>()
+                {
+                    {"api-version", "1.0"}
+                }
+            });
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler()
+            {
+                Method = HttpMethod.Post,
+                ResponseMessage = MockHelpers.CreateSuccessTokenResponseMessage(TestConstants.DefaultUniqueId, "user2@id.com", TestConstants.DefaultResource),
+                PostData = new Dictionary<string, string>()
+                {
+                    {"grant_type", "password"},
+                    {"username", "user2@id.com"},
+                    {"password", TestConstants.DefaultPassword},
+                }
+            });
+
+            TokenCache cache = new TokenCache();
+
+            var context = new AuthenticationContext(TestConstants.DefaultAuthorityHomeTenant, true, new TokenCache());
+
+            TokenCacheKey key = new TokenCacheKey(TestConstants.DefaultAuthorityHomeTenant,
+            TestConstants.DefaultResource, TestConstants.DefaultClientId, TokenSubjectType.User,
+            TestConstants.DefaultUniqueId, TestConstants.DefaultDisplayableId);
+            var setupResult = new AuthenticationResultEx
+            {
+                RefreshToken = "some-rt",
+                ResourceInResponse = TestConstants.DefaultResource,
+                Result = new AuthenticationResult("Bearer", "existing-access-token",
+                    DateTimeOffset.UtcNow + +TimeSpan.FromMinutes(100))
+            };
+
+            setupResult.Result.UserInfo = new UserInfo();
+            setupResult.Result.UserInfo.DisplayableId = TestConstants.DefaultDisplayableId;
+            context.TokenCache.tokenCacheDictionary[key] = setupResult;
+
+            var result =
+                await
+                    context.AcquireTokenAsync(TestConstants.DefaultResource, TestConstants.DefaultClientId,
+                        new UserPasswordCredential("user2@id.com", TestConstants.DefaultPassword));
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual(TestConstants.DefaultAuthorityHomeTenant, context.Authenticator.Authority);
+            Assert.AreEqual(result.AccessToken, "some-access-token");
+            Assert.IsNotNull(result.UserInfo);
+            Assert.AreEqual("user2@id.com", result.UserInfo.DisplayableId);
+            Assert.AreEqual(TestConstants.DefaultUniqueId, result.UserInfo.UniqueId);
+            Assert.AreNotEqual(context.TokenCache.tokenCacheDictionary.Values.First(), context.TokenCache.tokenCacheDictionary.Values.Last());
+            // There should be only two cache entrys.
+            Assert.AreEqual(2, context.TokenCache.Count);
```

Comment:
add check to ensure 2 cache entries are for 2 different users. make sure  you check the key and the value.